File: Backend/src/main/java/lan/dk/podcastserver/utils/custom/ffmpeg/CustomRunProcessFunc.java
package lan.dk.podcastserver.utils.custom.ffmpeg;

import javaslang.collection.List;
import net.bramp.ffmpeg.RunProcessFunction;

import java.io.IOException;


/**
 * Created by kevin on 24/07/2016.
 */
public class CustomRunProcessFunc extends RunProcessFunction {

    private List<ProcessListener> listeners = List.empty();

    @Override
    public Process run(java.util.List<String> args) throws IOException {
        Process p = super.run(args);

        this.listeners = listeners.remove(
            listeners
                    .find(pl -> args.contains(pl.url()))
                    .map(pl -> pl.process(p))
                    .getOrElse(ProcessListener.DEFAULT_PROCESS_LISTENER)
        );

        return p;
    }

    public CustomRunProcessFunc add(ProcessListener pl) {
        this.listeners = listeners.append(pl);
        return this;
    }
}

File: backend/src/main/kotlin/com/github/davinkevin/podcastserver/utils/custom/ffmpeg/CustomRunProcessFunc.kt
package com.github.davinkevin.podcastserver.utils.custom.ffmpeg

import arrow.core.getOrElse
import arrow.syntax.collections.firstOption
import com.github.davinkevin.podcastserver.utils.custom.ffmpeg.ProcessListener.Companion.DEFAULT_PROCESS_LISTENER
import net.bramp.ffmpeg.RunProcessFunction
import java.io.IOException

/**
 * Created by kevin on 24/07/2016.
 */
class CustomRunProcessFunc(private var listeners: List<ProcessListener> = listOf()) : RunProcessFunction() {

    @Throws(IOException::class)
    override fun run(args: List<String>): Process {
        val p = super.run(args)

        val toBeRemoved = listeners
                .firstOption { pl -> args.contains(pl.url) }
                .map { it.withProcess(p) }
                .getOrElse { DEFAULT_PROCESS_LISTENER }

        this.listeners = listeners - toBeRemoved

        return p
    }

    fun add(pl: ProcessListener): CustomRunProcessFunc {
        this.listeners = listeners + pl
        return this
    }

    operator fun plus(pl: ProcessListener): CustomRunProcessFunc = this.add(pl)
}
